chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/ContourDataSetCache.java

```diff
@@ -0,0 +1,460 @@
+    protected static DataRange computeLocalRange(final double[] input, final int srcWidth, final int srcHeight,
+            final boolean computeLocalRange) {
+        final DataRange zDataRange = new DataRange();
+        if (computeLocalRange) {
+            final int length = srcWidth * srcHeight;
+            for (int i = 0; i < length; i++) {
+                zDataRange.add(input[i]);
+            }
+        }
+        return zDataRange;
+    }
+
+    protected static void copySubFrame(final DataSet dataSet, final double[] dataBuffer,
+            final boolean parallelImplementation, //
+            final boolean xInverted, final int xMinIndex, final int xMaxIndex, //
+            final boolean yInverted, final int yMinIndex, final int yMaxIndex) {
+        final int width = Math.abs(xMaxIndex - xMinIndex) + 1;
+        final int height = Math.abs(yMaxIndex - yMinIndex) + 1;
+        final int dataLength = width * height;
+
+        if (!parallelImplementation) {
+            computeCoordinates(dataSet, dataBuffer, dataLength, xInverted, xMinIndex, xMaxIndex, yInverted, yMinIndex,
+                    yMaxIndex, yMinIndex);
+            return;
+        }
+
+        final int nMaxThreads = CachedDaemonThreadFactory.getNumbersOfThreads();
+        final int minthreshold = REF_HEIGHT_PARALLEL / 2; // TODO: tune this limit
+        final int divThread = (int) Math.ceil(height / (double) nMaxThreads);
+        final int stepSize = Math.max(divThread, minthreshold);
+        final List<Callable<Boolean>> workers = new ArrayList<>();
+        for (int i = yMinIndex; i < yMaxIndex; i += stepSize) {
+            final int start = i;
+            workers.add(() -> {
+                final int yMinLocal = start;
+                final int yMaxLocal = Math.min(start + stepSize, yMaxIndex);
+                computeCoordinates(dataSet, dataBuffer, dataLength, //
+                        xInverted, xMinIndex, xMaxIndex, //
+                        yInverted, yMinLocal, yMaxLocal, //
+                        yMinIndex);
+                return Boolean.TRUE;
+            });
+        }
+
+        try {
+            final List<Future<Boolean>> jobs = CachedDaemonThreadFactory.getCommonPool().invokeAll(workers);
+            for (final Future<Boolean> future : jobs) {
+                final Boolean r = future.get();
+                if (Boolean.FALSE.equals(r)) {
+                    throw new IllegalStateException(PARALLEL_WORKER_ERROR);
+                }
+            }
+        } catch (final InterruptedException | ExecutionException e) {
+            throw new IllegalStateException(PARALLEL_WORKER_ERROR, e);
+        }
+    }
+
+    protected static double quantize(final double value, final int nLevels) {
+        return ((int) (value * nLevels)) / (double) nLevels;
+        // original: return Math.round(value * nLevels) / (double) nLevels;
+    }
+
+    protected static double[] reduceData(final double[] inputBuffer, int inputWidth, int inputHeight,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'reduceData(double, int, int, double, int, int, int, ReductionType, boolean)' has an NPath complexity of 433](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=4880492)